targets/TARGET_Realtek/TARGET_AMEBA/spi_api.c

```diff
@@ -0,0 +1,1102 @@
+    uint32_t ssi_mosi, ssi_miso, ssi_peri;
+    uint8_t  ssi_idx, ssi_pinmux;
+    PHAL_SSI_ADAPTOR pHalSsiAdaptor;
+    PHAL_SSI_OP pHalSsiOp;
+
+    _memset((void*)obj, 0, sizeof(spi_t));
+    obj->state = 0;
+    uint32_t SystemClock = SystemGetCpuClk();
+    uint32_t MaxSsiFreq  = (SystemClock >> 2) >> 1;
+
+    /* SsiClockDivider doesn't support odd number */
+
+    DBG_SSI_INFO("SystemClock: %d\n", SystemClock);
+    DBG_SSI_INFO("MaxSsiFreq : %d\n", MaxSsiFreq);
+
+    ssi_mosi = pinmap_peripheral(mosi, PinMap_SSI_MOSI);
+    ssi_miso = pinmap_peripheral(miso, PinMap_SSI_MISO);
+
+    ssi_peri = pinmap_merge(ssi_mosi, ssi_miso);
+    if (unlikely(ssi_peri == NC)) {
+        DBG_SSI_ERR("spi_init(): Cannot find matched SSI index.\n");
+        return;
+    }
+    obj->sclk = (u8)sclk;
+    ssi_idx = RTL_GET_PERI_IDX(ssi_peri);
+    ssi_pinmux = RTL_GET_PERI_SEL(ssi_peri);
+    DBG_SSI_INFO("ssi_peri: %d, ssi_idx: %d, ssi_pinmux: %d\n", ssi_peri, ssi_idx, ssi_pinmux);
+
+    pHalSsiAdaptor = &obj->spi_adp;
+    pHalSsiOp = &obj->spi_op;
+
+    pHalSsiAdaptor->Index = ssi_idx;
+    pHalSsiAdaptor->PinmuxSelect = ssi_pinmux;
+    pHalSsiAdaptor->Role = SSI_MASTER;
+
+    HalSsiOpInit((VOID*)pHalSsiOp);
+
+    pHalSsiOp->HalSsiSetDeviceRole(pHalSsiAdaptor, pHalSsiAdaptor->Role);
+
+    /* Pinmux workaround */
+    if ((ssi_idx == 0) && (ssi_pinmux == SSI0_MUX_TO_GPIOC)) {
+        EEPROM_PIN_CTRL(OFF);
+    }
+
+    if ((ssi_idx == 0) && (ssi_pinmux == SSI0_MUX_TO_GPIOE)) {
+            DBG_SSI_WARN(ANSI_COLOR_MAGENTA"SPI0 Pin may conflict with JTAG\r\n"ANSI_COLOR_RESET);        
+    }
+
+    //TODO: Implement default setting structure.
+    pHalSsiOp->HalSsiLoadSetting(pHalSsiAdaptor, (void*)&SpiDefaultSetting);
+    pHalSsiAdaptor->DefaultRxThresholdLevel = SpiDefaultSetting.RxThresholdLevel;
+
+    if(HalSsiInit(pHalSsiAdaptor) != HAL_OK){
+        DBG_SSI_ERR(ANSI_COLOR_RED"spi_init(): SPI %x init fails.\n"ANSI_COLOR_RESET,pHalSsiAdaptor->Index);
+        return;        
+    }
+
+#ifndef CONFIG_MBED_ENABLED  
+    pHalSsiAdaptor->TxCompCallback = spi_tx_done_callback;
+    pHalSsiAdaptor->TxCompCbPara = (void*)obj;
+    pHalSsiAdaptor->RxCompCallback = spi_rx_done_callback;
+    pHalSsiAdaptor->RxCompCbPara = (void*)obj;
+    pHalSsiAdaptor->TxIdleCallback = spi_bus_tx_done_callback;
+    pHalSsiAdaptor->TxIdleCbPara = (void*)obj;
+
+#ifdef CONFIG_GDMA_EN    
+    HalGdmaOpInit((VOID*)&SpiGdmaOp);
+    pHalSsiAdaptor->DmaConfig.pHalGdmaOp = &SpiGdmaOp;
+    pHalSsiAdaptor->DmaConfig.pRxHalGdmaAdapter = &obj->spi_gdma_adp_rx;
+    pHalSsiAdaptor->DmaConfig.pTxHalGdmaAdapter = &obj->spi_gdma_adp_tx;
+    obj->dma_en = 0;
+    pHalSsiAdaptor->HaveTxChannel = 0;
+    pHalSsiAdaptor->HaveRxChannel = 0;
+#endif
+#endif
+	osDelay(1);
+}
+
+void spi_free (spi_t *obj)
+{
+    PHAL_SSI_ADAPTOR pHalSsiAdaptor;    
+    pHalSsiAdaptor = &obj->spi_adp;
+    HalSsiDeInit(pHalSsiAdaptor);
+
+    SPI0_MULTI_CS_CTRL(OFF);
+#ifndef CONFIG_MBED_ENABLED    
```

Comment:
@0xc0170 I think @Archcady has pushed two more commits to remove redundant macros and dead code.